<div style="background:linear-gradient(135deg,#1e1b4b 0%,#4338ca 55%,#6366f1 100%);border-radius:18px;padding:34px 30px;color:#fff;font-family:Inter,Segoe UI,sans-serif">
  <div style="font-size:12px;letter-spacing:3px;color:#c7d2fe;font-weight:700;text-transform:uppercase">Chapter 30 · Solutions</div>
  <div style="font-size:34px;font-weight:900;line-height:1.1;margin:10px 0 6px">Probability Challenges, Worked Answers ✅</div>
  <div style="font-size:15px;color:#e0e7ff;max-width:720px;line-height:1.6">Full solutions to the five Probability Fundamentals challenges. Try them yourself first, then compare.</div>
  <div style="margin-top:16px;font-size:13px;color:#cbd5e1">Statistics, Data Science and AI: A Visual Handbook · John Fisher · 2026</div>
</div>

## ⚙️ Setup

In [1]:
import numpy as np, pandas as pd, matplotlib.pyplot as plt
from itertools import product
rng = np.random.default_rng(303)
BLUE="#2563eb"; INDIGO="#4f46e5"; GREEN="#059669"; PINK="#db2777"; INK="#1a2138"; GRID="#e6e9f2"
plt.rcParams.update({"figure.dpi":110,"axes.grid":True,"grid.color":GRID,"axes.axisbelow":True,
   "axes.spines.top":False,"axes.spines.right":False,"axes.titleweight":"bold"})
print("ready")

ready


<div style="background:#eef2ff;border-left:5px solid #4f46e5;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#4f46e5;letter-spacing:1px">CHALLENGE 1 · SAMPLE SPACE</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">Two coins</div>
<div style="color:#4a5578;margin-top:6px">Flip a fair coin twice. Write the sample space, then list the outcomes in the events A = 'exactly one head' and B = 'at least one head'.</div>
</div>

In [2]:
S = list(product(["H","T"], repeat=2))
A = [o for o in S if o.count("H")==1]
B = [o for o in S if o.count("H")>=1]
print("sample space S:", S, " size", len(S))
print("A (exactly one head):", A)
print("B (at least one head):", B)

sample space S: [('H', 'H'), ('H', 'T'), ('T', 'H'), ('T', 'T')]  size 4
A (exactly one head): [('H', 'T'), ('T', 'H')]
B (at least one head): [('H', 'H'), ('H', 'T'), ('T', 'H')]


**Answer:** the sample space has 4 equally likely outcomes: HH, HT, TH, TT. "Exactly one head" is {HT, TH} (2 outcomes); "at least one head" is {HH, HT, TH} (3 outcomes). Listing the sample space first makes every later probability a matter of counting the right subset.

<div style="background:#eef2ff;border-left:5px solid #4f46e5;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#2563eb;letter-spacing:1px">CHALLENGE 2 · CLASSICAL</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">Probability from counting</div>
<div style="color:#4a5578;margin-top:6px">Using that two-coin sample space, compute P(exactly one head) and P(at least one head) classically (favorable / total).</div>
</div>

In [3]:
S = list(product(["H","T"], repeat=2))
p_one = sum(o.count("H")==1 for o in S)/len(S)
p_atleast = sum(o.count("H")>=1 for o in S)/len(S)
print(f"P(exactly one head)  = 2/4 = {p_one:.2f}")
print(f"P(at least one head) = 3/4 = {p_atleast:.2f}")

P(exactly one head)  = 2/4 = 0.50
P(at least one head) = 3/4 = 0.75


**Answer:** P(exactly one head) = 2/4 = 0.50 and P(at least one head) = 3/4 = 0.75. Because the four outcomes are equally likely, classical probability is just favorable-over-total, no experiment required (Laplace's definition).

<div style="background:#eef2ff;border-left:5px solid #4f46e5;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#059669;letter-spacing:1px">CHALLENGE 3 · EMPIRICAL</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">Simulate and compare</div>
<div style="color:#4a5578;margin-top:6px">Estimate P(at least one head in two flips) by simulating 50,000 experiments, and compare to the theoretical 0.75.</div>
</div>

In [4]:
sims = rng.integers(0,2,(50_000,2))     # 1 = heads
emp = np.mean(sims.sum(axis=1) >= 1)
print(f"empirical P(at least one head) = {emp:.3f}")
print(f"theoretical                    = 0.750")
print(f"difference                     = {abs(emp-0.75):.3f}")

empirical P(at least one head) = 0.750
theoretical                    = 0.750
difference                     = 0.000


**Answer:** the simulated proportion lands within a few thousandths of 0.75. Empirical probability (long-run relative frequency) and classical probability agree because of the Law of Large Numbers, with 50,000 trials the estimate is already very tight.

<div style="background:#eef2ff;border-left:5px solid #4f46e5;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#d97706;letter-spacing:1px">CHALLENGE 4 · COMPLEMENT</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">The easy way in</div>
<div style="color:#4a5578;margin-top:6px">A die is rolled three times. Find P(at least one six). Do it with the complement rule, then verify by simulation.</div>
</div>

In [5]:
p = 1 - (5/6)**3
print(f"P(no six in 3 rolls) = (5/6)^3 = {(5/6)**3:.3f}")
print(f"P(at least one six)  = 1 - that = {p:.3f}")
sims = rng.integers(1,7,(200_000,3))
emp = np.mean((sims==6).any(axis=1))
print(f"empirical (200k)     = {emp:.3f}")

P(no six in 3 rolls) = (5/6)^3 = 0.579
P(at least one six)  = 1 - that = 0.421
empirical (200k)     = 0.424


**Answer:** "at least one six" is painful to count directly (one six, two sixes, three sixes), but its complement "no six at all" is easy: (5/6)^3. So P(at least one six) = 1 - (5/6)^3 ≈ 0.421, confirmed by simulation. Whenever you see "at least one," reach for the complement.

<div style="background:#eef2ff;border-left:5px solid #4f46e5;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#db2777;letter-spacing:1px">CHALLENGE 5 · INDEPENDENCE</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">No such thing as 'due'</div>
<div style="color:#4a5578;margin-top:6px">Simulate a fair coin. After every run of four tails in a row, what fraction of the very next flips are heads? Does a long tail streak make heads 'due'?</div>
</div>

In [6]:
flips = rng.integers(0,2,300_000)       # 1 = heads, 0 = tails
nxt = []; streak = 0
for i in range(len(flips)-1):
    streak = streak+1 if flips[i]==0 else 0
    if streak >= 4: nxt.append(flips[i+1])
nxt = np.array(nxt)
print(f"flips after a run of >=4 tails : {len(nxt):,}")
print(f"fraction that were heads       : {nxt.mean():.3f}")
print("theory: 0.500 -- heads are never \"due\".")

flips after a run of >=4 tails : 18,814
fraction that were heads       : 0.499
theory: 0.500 -- heads are never "due".


**Answer:** about half. A streak of tails does not make heads any more likely on the next flip, the coin has no memory, and the flips are independent. Believing otherwise is the gambler's fallacy. (Note: this is different from saying a *long run* will average out, which it does; it is the *next single flip* that is unaffected.)

---
<div style="background:#ffffff;border:1px solid #e6e9f2;border-radius:16px;padding:22px 26px;font-family:Inter,sans-serif">
<div style="font-size:19px;font-weight:800;color:#1a2138">🎉 Nicely done</div>
<div style="color:#4a5578;line-height:1.8;margin-top:8px">You built sample spaces, computed probability classically and empirically, used the complement as a shortcut, and saw why independent events have no memory. These fundamentals power every rule, distribution, and inference still to come.</div>
</div>

<div style="text-align:center;color:#8b94b3;font-size:12px;margin-top:16px">Statistics, Data Science and AI: A Visual Handbook · © 2026 John Fisher</div>